### PR TITLE
Heart rate request handler

### PR DESCRIPTION
**Pull request by tiginamaria**

### CODE REVIEW

File: server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java [hunk 1]
```diff
@@ -4,18 +4,64 @@
...
+
+    @Override
+    protected void onPause() {
+        mBluetoothServer.unregisterReceiver();
+        super.onPause();
+    }
+
+    @Override
+    protected void onDestroy() {
+        mBluetoothServer.stop();
```
Comment (ijuhoor) [inline server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java:59]: Maybe the heart rate sensor listener need to be unregistered here.

File: server/app/src/main/java/com/google/heartrate/wearos/app/sensors/HeartRateSensorListener.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        Log.v(TAG, String.format("onSensorChanged() - value=%d", heartRate));
+
+        for (HeartRateValueSubscriber listener : subscribers) {
+            listener.onHeartRateValueChanged(heartRate);
+        }
+    }
+
+    @Override
+    public void onAccuracyChanged(Sensor sensor, int accuracy) {
+
```
Comment (ijuhoor) [inline server/app/src/main/java/com/google/heartrate/wearos/app/sensors/HeartRateSensorListener.java:96]: Maybe log that this is not supported.

File: server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/BluetoothServerCallback.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        Log.v(TAG, String.format("onCharacteristicWriteRequest() - device=%s characteristic=%s value=%s",
+                characteristic.getUuid(),
+                Arrays.toString(characteristic.getValue()),
+                Arrays.toString(value)));
+        try {
+            GattServiceRequestHandler requestHandler = mBluetoothServer.getGattServiceRequestHandler(characteristic);
+            requestHandler.onCharacteristicWrite(device, characteristic, offset, value);
+
+            if (responseNeeded) {
+                mBluetoothServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value);
```
Comment (ijuhoor) [inline server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/BluetoothServerCallback.java:131]: Totally optional:
 It might be good to log when `sendResponse` fails:
```
if (!mBluetoothServer.sendResponse( ... )) {
  Log.e(TAG, ....)
}
```

File: server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java [hunk 4]
```diff
@@ -4,18 +4,81 @@
...
-        setAmbientEnabled();
+    /**
+     * Start server with heart rate service hosted.
+     */
+    private void startHeartRateServer() {
+        try {
+            heartRateSensorListener = new HeartRateSensorListener(this);
+            heartRateSensorListener.registerSubscriber(this);
+
+            bluetoothServer = new BluetoothServer(this);
```
Comment (sanderalewijnse) [inline server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java:49]: Same for the bluetooth server, why not construct it in the service?

File: server/app/src/main/java/com/google/heartrate/wearos/app/sensors/HeartRateSensorListener.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+
+/**
+ * Listener for heart rate sensor.
+ */
+public class HeartRateSensorListener implements SensorEventListener {
+    private static final String TAG = HeartRateSensorListener.class.getSimpleName();
+
+    /** Sensor manager. */
+    private final SensorManager mSensorManager;
```
Comment (sanderalewijnse) [inline server/app/src/main/java/com/google/heartrate/wearos/app/sensors/HeartRateSensorListener.java:20]: Let's be consistent with the rest of your code: mSensorManager -> sensorManager?

File: server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/handlers/HeartRateServiceRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            Log.d(TAG, String.format("Subscribe device %s to notifications", device));
+            registeredDeviceStorage.addDevice(device);
+        } else if (Arrays.equals(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE, value)) {
+            Log.d(TAG, String.format("Unsubscribe device %s from notifications", device));
+            registeredDeviceStorage.removeDevice(device);
+        }
+    }
+
+    /**
+     * Callback to listen to {@link HeartRateSensorListener} about heart rate value change.
```
Comment (sanderalewijnse) [inline server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/handlers/HeartRateServiceRequestHandler.java:119]: I'd rephrase this line a bit. Maybe: "Called when the heart rate value was changed"?

File: server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java [hunk 7]
```diff
@@ -4,18 +4,81 @@
...
+    }
 
-        // Enables Always-on
-        setAmbientEnabled();
+    /**
+     * Start server with heart rate service hosted.
+     */
+    private void startHeartRateServer() {
+        try {
+            heartRateSensorListener = new HeartRateSensorListener(this);
```
[the remark this answers is not in the document]
Comment (tiginamaria) [reply server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java:46]: See next PR

File: server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java (hunk 4 above)
Comment (tiginamaria) [reply server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java:49]: See next PR
